src/CodeViewer/Viewer.cpp

```diff
@@ -128,47 +192,156 @@ void Viewer::DrawLineNumbers(QPaintEvent* event) {
   }
 }
 
-void Viewer::UpdateLeftSidebarWidth() {
+// For example, from a value = 0.5, we want to print "50.00 %"
+static QString DoubleToPercentageString(double v) {
+  v = std::clamp(v, 0., 1.);
+  return QString{"%1 %"}.arg(v * 100., 0, 'f', 2);
+}
+
+void Viewer::DrawSampleCounters(QPaintEvent* event) {
+  if (!sample_counters_enabled_) return;
+  QPainter painter{&right_sidebar_widget_};
+  painter.setFont(font());
+  painter.fillRect(event->rect(), kLineNumberBackgroundColor);
+
+  const auto top_of = [&](const QTextBlock& block) {
+    return static_cast<int>(
+        std::round(blockBoundingGeometry(block).translated(contentOffset()).top()));
+  };
+  const auto bottom_of = [&](const QTextBlock& block) {
+    return top_of(block) + static_cast<int>(std::round(blockBoundingRect(block).height()));
+  };
+
+  for (auto block = firstVisibleBlock(); block.isValid() && top_of(block) <= event->rect().bottom();
+       block = block.next()) {
+    if (!block.isVisible() || bottom_of(block) < event->rect().top()) continue;
+
+    const int left = left_margin_.ToPixels(fontMetrics());
+
+    auto opt_num_samples_in_lines = code_report_->GetNumSamplesAtLine(block.blockNumber() + 1);
+    if (opt_num_samples_in_lines.has_value()) {
```

Comment:
Nit: `if (!opt_num_samples_in_lines.has_value()) return;`

Reply:
I wasn't sure there, but I will change it!

Reply:
BTW, I should write _continue_ instead, not return